### PR TITLE
Vendor in latest container/storage for devicemapper support

### PR DESCRIPTION
**Pull request by rhatdan**
container/storage now supports devicemapper options that allow you to
configure it.  This might be useful to katacontainers.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM, kata-jenkins still isn't hip though.

**Comment by wking** [comment]
Not something that needs to block this PR, but with the configuration options still developing, we may want to mention which version of container/storage we're using [here][1], were users will be able to easily discover it.  Although by using a vendored package (vs. a system package), it's not clear to me who will be in charge of installing the system `storage.conf(5)`.  It looks like we're currently assuming that another package will install the `storage.conf(5)` that corresponds to the version of container/storage we vendor.

[1]: https://github.com/kubernetes-incubator/cri-o/blame/6d98a70f1fd05d1ed4183a5e9dd36e0811008ad1/docs/crio.8.md#L157

**Comment by rhatdan** [comment]
@mrunalp @sboeuf Does Kata Containers testing run with devicemapper for storage on loopback devices.  The current code in containers/storage is turning off loopback support.  We need to either specify an actual device for your system to run your tests on, or figure out a way to pass Test=1 into containers storage to allow loopback support.

**Comment by sboeuf** [comment]
@chavafg could you answer @rhatdan about the storage driver used for Kata Containers testing ?

**Comment by sboeuf** [comment]
I guess we can but we would have to improve the script to create other devices than simple loopback ones. @amshinde @chavafg is that something you can look into ?

/cc @mcastelino

**Comment by rhatdan** [comment]
Well actually I am working on a change to containers-storage to allow you to modify it.  So we might have to disable this check to allow us to get the fix in, and then you could reenable it using a real device.

**Comment by sboeuf** [comment]
@rhatdan nice ! But I am actually thinking it would be also good for us to have support for real devices. How does CRI-O test with real devices in case of runc ?

**Comment by amshinde** [comment]
@chavafg We used to run tests using a real device in clear-containers. See these storage options:
https://github.com/clearcontainers/tests/blob/3c3bed5046fb460a33b98d3402da20e2024a13ad/Azure/cloud-init.txt#L197

Basically, 
```
export STORAGE_OPTIONS="--storage-driver devicemapper --storage-opt dm.directlvm_device=/dev/sdd, dm.thinp_percent=95, dm.thinp_metapercent=1, dm.thinp_autoextend_threshold=80, dm.thinp_autoextend_percent=20, dm.directlvm_device_force=true"
```

**Comment by sboeuf** [comment]
@amshinde thanks, this would be easier to actually avoid having to workaround containers/storage.
@chavafg let us know if you can make this work ;)

**Comment by chavafg** [comment]
sure, will test and open a PR if everything goes well with that options.

**Comment by rhatdan** [comment]
My patch will allow you to put all of that configuration data into storage.conf rather then having to understand the underlying layout.  But you could do that now.

**Comment by sboeuf** [comment]
That's cool :), we'll have the choice ! Thank you @rhatdan

**Comment by chavafg** [comment]
I created  `/dev/sdb` device in my host, and passed to the `STORAGE_OPTIONS`
```
export STORAGE_OPTIONS="--storage-driver devicemapper --storage-opt dm.use_deferred_removal=false, dm.directlvm_device=dev/sdb1 dm.directlvm_device_force=true"
```
But, now I got: 

```
not ok 1 ctr not found correct error message
# (from function `start_crio' in file helpers.bash, line 225,
#  in test file ctr.bats, line 10)
#   `start_crio' failed
# time="2018-05-21T21:36:26Z" level=error msg="must set --root and --runroot, or neither" 
# 2018/05/21 21:36:26 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/tmp.31mUjsIKJt/crio.sock: connect: no such file or directory"; Reconnect
ing to {/tmp/tmp.31mUjsIKJt/crio.sock <nil>}
# time="2018-05-21T21:36:26Z" level=fatal msg="listing containers failed: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
# 2018/05/21 21:36:26 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/tmp.31mUjsIKJt/crio.sock: connect: no such file or directory"; Reconnect
ing to {/tmp/tmp.31mUjsIKJt/crio.sock <nil>}
# time="2018-05-21T21:36:26Z" level=fatal msg="listing pod sandboxes failed: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
```

any idea?

**Comment by amshinde** [comment]
@chavafg Looks like a malformed option line. Try this : 
```sed -i 's/storage_option = \[/storage_option = \["dm.directlvm_device=\/dev\/sdb", "dm.thinp_percent=95", "dm.thinp_metapercent=1", "dm.thinp_autoextend_threshold=80", "dm.thinp_autoextend_percent=20", "dm.directlvm_device_force=true"/' "/etc/crio/crio.conf"```

**Comment by sboeuf** [comment]
@chavafg any chance you re-tried based on @amshinde comment ?

**Comment by chavafg** [comment]
@sboeuf @amshinde what about if I change our tests to run with overlay so that we can unblock this and when we figure out how to run with a block device I can change it again?

**Comment by chavafg** [comment]
I was able to start crio and run the first test successfully. After that, the cri-o tests are failing because they ask for another 'clean' block device. 
```
1..32
ok 1 ctr not found correct error message
not ok 2 ctr termination reason Completed
# (from function `start_crio' in file helpers.bash, line 225,
#  in test file ctr.bats, line 19)
#   `start_crio' failed
# time="2018-05-22T15:35:02Z" level=error msg="error opening storage: /dev/sdc is already part of a volume group "storage": must remove this device from any volume group or provide a different device" 
# 2018/05/22 15:35:02 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/tmp.utctYelsKY/crio.sock: connect: no such file or directory"; Reconnect
ing to {/tmp/tmp.utctYelsKY/crio.sock <nil>}
# time="2018-05-22T15:35:02Z" level=fatal msg="listing containers failed: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
# 2018/05/22 15:35:02 grpc: addrConn.resetTransport failed to create client transport: connection error: desc = "transport: dial unix /tmp/tmp.utctYelsKY/crio.sock: connect: no such file or directory"; Reconnect
ing to {/tmp/tmp.utctYelsKY/crio.sock <nil>}
# time="2018-05-22T15:35:02Z" level=fatal msg="listing pod sandboxes failed: rpc error: code = Unavailable desc = grpc: the connection is unavailable"
# 0
```

I am providing these `STORAGE_OPTIONS:`
```
export STORAGE_OPTIONS="--storage-driver devicemapper --storage-opt dm.directlvm_device=/dev/sdc --storage-opt dm.directlvm_device_force=true --storage-opt dm.thinp_percent=95 --storage-opt dm.thinp_metapercent=1 --storage-opt dm.thinp_autoextend_threshold=80 --storage-opt dm.thinp_autoextend_percent=20"
```

**Comment by chavafg** [comment]
any idea of how can I make the whole tests work using the same device?

**Comment by rhatdan** [comment]
Try dm.directlvm_device_force=True

**Comment by chavafg** [comment]
Opened #1576 to see if someone can help us with the `bin2img` failure.

**Comment by sboeuf** [comment]
@chavafg any progress there ?
If not, we might need to consider running overlay testing for CRI-O Jenkins job (as you suggested @chavafg ) until we figure out this properly. But I don't want to lower the priority on this and forget that we're not testing on devicemapper anymore. That's why it'd be cool to be able to solve this instead of switching.

/cc @egernst

**Comment by rhatdan** [comment]
@chavafg @sameo I have added a cleanup_lvm section to the test suite which will be called on each cleanup. If you set the LVM_DEVICE environment variable to the device you are using, this should execute @nalind suggested fix.  If this passes all of the tests other then Kata could you guys add it to your kata tests so that we can finally get this PR merged.

**Comment by chavafg** [comment]
great, thanks @rhatdan, I need to add the `$LVM_DEVICE` variable in our test script, will add the fix right away and let you know when you can restart the kata jobs and hopefully everything willl work now.

File: test/helpers.bash [hunk 1]
```diff
@@ -341,10 +341,19 @@ function restart_crio() {
...
+		lvm vgremove -y storage
+		lvm pvremove -y $LVM_DEVICE
+	fi
+}
+
 function cleanup_test() {
 	cleanup_ctrs
 	cleanup_pods
 	stop_crio
+	cleaup_lvm
```
Comment (chavafg) [inline test/helpers.bash]: typo on `cleanup_lvm`

**Comment by rhatdan** [comment]
Fixed

**Comment by chavafg** [comment]
@rhatdan can you restart the kata job? Now everything should be fine.

Thanks

**Comment by chavafg** [comment]
\o/ kata worked now :)

**Comment by rhatdan** [comment]
@mrunalp @runcom @wking @giuseppe PTAL

**Comment by rhatdan** [comment]
@chavafg Nice work.

**Comment by sboeuf** [comment]
Thanks @chavafg, @rhatdan and @nalind for solving this ;)

**Comment by giuseppe** [comment]
@rhatdan LGTM

**Comment by runcom** [comment]
@mrunalp PTAL I believe this can be merged to also unblock PRs failing at kata-containers job

**Comment by rhatdan** [comment]
@runcom Merging.

**Pull request merged**